pkg/types/ruletypes/evaluation.go

```diff
@@ -0,0 +1,285 @@
+		}
+		if *cs.Minute < 0 || *cs.Minute > 59 {
+			return errors.NewInvalidInputf(errors.CodeInvalidInput, "minute must be between 0 and 59")
+		}
+	case ScheduleTypeMonthly:
+		if cs.Day == nil || cs.Hour == nil || cs.Minute == nil {
+			return errors.NewInvalidInputf(errors.CodeInvalidInput, "day, hour and minute must be specified for monthly schedule")
+		}
+		if *cs.Day < 1 || *cs.Day > 31 {
+			return errors.NewInvalidInputf(errors.CodeInvalidInput, "day must be between 1 and 31")
+		}
+		if *cs.Hour < 0 || *cs.Hour > 23 {
+			return errors.NewInvalidInputf(errors.CodeInvalidInput, "hour must be between 0 and 23")
+		}
+		if *cs.Minute < 0 || *cs.Minute > 59 {
+			return errors.NewInvalidInputf(errors.CodeInvalidInput, "minute must be between 0 and 59")
+		}
+	default:
+		return errors.NewInvalidInputf(errors.CodeInvalidInput, "invalid schedule type")
+	}
+	return nil
+}
+
+func (cumulativeWindow CumulativeWindow) NextWindowFor(curr time.Time) (time.Time, time.Time) {
+	loc := time.UTC
+	if cumulativeWindow.Timezone != "" {
+		if tz, err := time.LoadLocation(cumulativeWindow.Timezone); err == nil {
+			loc = tz
+		}
+	}
+
+	currInTZ := curr.In(loc)
+	windowStart := cumulativeWindow.getLastScheduleTime(currInTZ, loc)
+
+	return windowStart.In(time.UTC), currInTZ.In(time.UTC)
+}
+
+func (cw CumulativeWindow) getLastScheduleTime(curr time.Time, loc *time.Location) time.Time {
+	schedule := cw.Schedule
+
+	switch schedule.Type {
+	case ScheduleTypeHourly:
+		// Find the most recent hour boundary with the specified minute
+		minute := *schedule.Minute
+		candidate := time.Date(curr.Year(), curr.Month(), curr.Day(), curr.Hour(), minute, 0, 0, loc)
+		if candidate.After(curr) {
+			candidate = candidate.Add(-time.Hour)
+		}
+		return candidate
+
+	case ScheduleTypeDaily:
+		// Find the most recent day boundary with the specified hour and minute
+		hour := *schedule.Hour
+		minute := *schedule.Minute
+		candidate := time.Date(curr.Year(), curr.Month(), curr.Day(), hour, minute, 0, 0, loc)
+		if candidate.After(curr) {
+			candidate = candidate.AddDate(0, 0, -1)
+		}
+		return candidate
+
+	case ScheduleTypeWeekly:
+		weekday := time.Weekday(*schedule.Weekday)
+		hour := *schedule.Hour
+		minute := *schedule.Minute
+
+		// Calculate days to subtract to reach the target weekday
+		daysBack := int(curr.Weekday() - weekday)
+		if daysBack < 0 {
+			daysBack += 7
+		}
+
+		candidate := time.Date(curr.Year(), curr.Month(), curr.Day(), hour, minute, 0, 0, loc).AddDate(0, 0, -daysBack)
+		if candidate.After(curr) {
+			candidate = candidate.AddDate(0, 0, -7)
+		}
+		return candidate
+
+	case ScheduleTypeMonthly:
+		// Find the most recent month boundary with the specified day, hour and minute
+		targetDay := *schedule.Day
+		hour := *schedule.Hour
+		minute := *schedule.Minute
+
+		// Try current month first
+		lastDayOfCurrentMonth := time.Date(curr.Year(), curr.Month()+1, 0, 0, 0, 0, 0, loc).Day()
```

Comment:
Why does it say `lastDayOfCurrentMonth` but uses `curr.Month()+1`? Isn't this a first day of the next month?

Reply:
Ok, i didn't know the time.Date worked that way.